api/cds.proto

```diff
@@ -31,86 +33,151 @@ service ClusterDiscoveryService {
+    // upstream cluster. If not specified, the default is 3.
     google.protobuf.UInt32Value max_retries = 5;
   }
+
+  // If multiple :ref:`Thresholds<envoy_api_msg_CircuitBreakers.Thresholds>`
+  // are defined with the same :ref:`RoutingPriority<envoy_api_enum_RoutingPriority>`,
+  // the first one in the list is used. If no Thresholds is defined for a given
+  // :ref:`RoutingPriority<envoy_api_enum_RoutingPriority>`, the default values
+  // are used.
   repeated Thresholds thresholds = 1;
 }
 
 message Cluster {
   // Supplies the name of the cluster which must be unique across all clusters.
-  // Any : in the cluster name will be converted to _ when emitting statistics.
-  // TODO(htuch): cross-link to
-  // https://www.envoyproxy.io/envoy/operations/cli#cmdoption-max-obj-name-len.
-  string name = 1;
-
-  // The service discovery type to use for resolving the cluster.
+  // The cluster name is used when emitting
+  // :ref:`statistics <config_cluster_manager_cluster_stats>`.
+  // Any ``:`` in the cluster name will be converted to ``_`` when emitting statistics.
+  // By default, the maximum length of a cluster name is limited to 60
+  // characters. This limit can be increased by setting the
+  // :option:`--max-obj-name-len` command line argument to the desired value.
+  string name = 1 [(validate.rules).string.min_len = 1];
+
+  // Refer to :ref:`service discovery type <arch_overview_service_discovery_types>`
+  // for an explanation on each type.
   enum DiscoveryType {
+    // Refer to the :ref:`static discovery type<arch_overview_service_discovery_types_static>`
+    // for an explanation.
     STATIC = 0;
+
+    // Refer to the :ref:`strict DNS discovery
+    // type<arch_overview_service_discovery_types_strict_dns>`
+    // for an explanation.
     STRICT_DNS = 1;
+
+    // Refer to the :ref:`logical DNS discovery
+    // type<arch_overview_service_discovery_types_logical_dns>`
+    // for an explanation.
     LOGICAL_DNS = 2;
+
+    // Refer to the :ref:`service discovery type<arch_overview_service_discovery_types_sds>`
+    // for an explanation.
     EDS = 3;
+
+    // Refer to the :ref:`original destination discovery
+    // type<arch_overview_service_discovery_types_original_destination>`
+    // for an explanation.
     ORIGINAL_DST = 4;
   }
-  DiscoveryType type = 2;
+  // The :ref:`service discovery type <arch_overview_service_discovery_types>`
+  // to use for resolving the cluster.
+  DiscoveryType type = 2 [(validate.rules).enum.defined_only = true];
 
   // Only valid when discovery type is EDS.
   message EdsClusterConfig {
+    // Configuration for the source of EDS updates for this Cluster.
     ConfigSource eds_config = 1;
+
     // Optional alternative to cluster name to present to EDS. This does not
     // have the same restrictions as cluster name, i.e. it may be arbritary
     // length.
-    string service_name = 2;
+    string service_name = 2 [(validate.rules).string.min_len = 1];
```

Comment:
qq: does min_len force it to be required?

Reply:
Yep.

Reply:
oops, merged before I got a chance to fix. I talked to @rodaine about this earlier in a different context, and wanted to delete this rule as the field is optional. @htuch can you fix this in your PR?

Reply:
Yes, will do.